Johannes/owned handle

jkneer opened the pull request:
- Added lifetimes, open point: should LvOwned be 'static?
- Implemented Clone on LvOwned (removed unsafe From impl)
- Implemented Borrow and ToOwned on UHandle
- added fn handle to LvOwned

untested

Adresses #20 , #21 , #22

JamesMc86 commented on labview-interop/src/memory.rs:
````diff
@@ -195,21 +228,21 @@ mod lv_owned {
     ///```
     #[repr(transparent)]
-    pub struct LvOwned<T: ?Sized>(UHandle<T>);
+    pub struct LvOwned<'a, T: ?Sized>(pub(crate) UHandle<'a, T>);
````
As we spoke about on the issue. Please make the inner UHandle 'static and then we don't need the lifetime on LvOwned which simplifies the rest of the Owned API.

jkneer replied:
This will take me more time.
Last time I added the lifetime annotation to UHandle and then went with the linter and added lifetimes everywhere needed to make the linter happy.

When I now try to make the LvOwned internal UHandle static I will need to refactor ToOwned and clone_... It needs to be clear that after cloning the lifetime has become static. ToOwned seems to support that, but I haven't been successful in expressing my intent in correct syntax...

JamesMc86 replied:
I understand. This is where the `new_unsized` method should help. That should be able to create the handle with the `'static` lifetime and then pass it to the initialisation function to do the copy.

The owned string shows an example of how to call this at https://github.com/WiresmithTech/Rust-LabVIEW-Interop/blob/50060a63374659bd34391db3cb5317b7760c93f1/labview-interop/src/types/string.rs#L230

jkneer replied:
That was not my problem :)

I either run into trait bounds, lifetimes or type incompatibilies in to_owned()... something seems to just line up.

JamesMc86 commented on labview-interop/src/memory.rs:
```diff
@@ -141,8 +143,32 @@ impl<T: ?Sized> DerefMut for UHandle<T> {
+
+//TODO test
+#[cfg(feature = "link")]
+impl<'a, T: ?Sized> ToOwned for UHandle<'a, T> {
```
This implementation should be in the lvowned submodule. Grouping it with LvOwned makes sense and I put it in a submodule so we can have the single feature flag for link to cover all of the contents.

JamesMc86 commented on labview-interop/src/memory.rs:
```diff
@@ -219,34 +252,58 @@ mod lv_owned {
+        fn clone(&self) -> Self {
+            let mut cloned_handle = UHandle(std::ptr::null_mut() as *mut *mut T, PhantomData);
+            unsafe {
+                self.clone_into_pointer(&mut cloned_handle as *mut UHandle<T>);
```
Happy to take as is for now and test if this works with a null handle in. 

However I may refactor this to use the existing new method and provide an init function to do the copy so we have a single implementation of setting up a new owned handle.

JamesMc86 reviewed (changes requested):
Again, happy to take a look at testing if you can make a couple of these changes. The main one being to remove the lifetime from LvOwned.

Thanks again for helping with this!

jkneer commented:
This will take me more time.
Last time I added the lifetime annotation to UHandle and then went with the linter and added lifetimes everywhere needed to make the linter happy. 

When I now try to make the LvOwned internal UHandle static I will need to refactor ToOwned and clone_... It needs to be clear that after cloning the lifetime has become static. ToOwned seems to support that, but I haven't been successful in expressing my intent in correct syntax...

jkneer commented:
Please have a look, I can't find a solution and have to give up for now. 

The current version needs to implement ?Sized + Sized to get rid of the incompatibilities in the ToOwned chain...

JamesMc86 commented:
No worries, I'll take a look at it in the morning but there are a lot of related changes here. Probably the easiest thing is I can land it as it is and then make the necessary changes to make the lifetimes work.

JamesMc86 reviewed (approved):
Will make lifetime changes after merge.
